core/src/impl/Kokkos_Half_FloatingPointWrapper.hpp

```diff
@@ -23,6 +23,18 @@
 #include <iosfwd>  // istream & ostream for extraction and insertion ops
 #include <string>
 
+namespace Kokkos::Experimental {
+/// @brief templated struct for determining if half_t is an alias to float.
+/// @tparam T The type to specialize on.
+template <class T>
+struct kokkos_type_is_half_t : std::false_type {};
+
+/// @brief templated struct for determining if bhalf_t is an alias to float.
+/// @tparam T The type to specialize on.
+template <class T>
+struct kokkos_type_is_bhalf_t : std::false_type {};
+}  // namespace Kokkos::Experimental
```

Comment:
Why introduce these traits in this PR?

The code below
```C++
static_assert(Kokkos::Experimental::kokkos_type_is_half_t<Kokkos::Experimental::half_t>::value);
```
would error out for some backends and not for others.
If you really want them you need to make the case why we might want them.

Reply:
I answered this above... See https://github.com/kokkos/kokkos/pull/6124#discussion_r1244191746

Reply:
I saw that and I don't find it convincing.  We can argue about it at the developer meeting tomorrow.

Reply:
So, T could be float in the math tests and the cleanest alternative I can think of to using the trait is to do this:
```
      constexpr bool const x_is_half = !std::is_same<T, float>::value && (std::is_same<T, KE::bhalf_t>::value || std::is_same<T, KE::half_t>::value);
```
I would have to test whether `std::is_same<T, float>::value` evalutes to true when T is KE::bhalf_t or T is KE::half_t and the half-type aliases to float. If it's not true, we would then need to use `#if defined(KOKKOS_HALF_T_IS_FLOAT) && !KOKKOS_HALF_T_IS_FLOAT`.

So, I can do the above but think everyone would find using the traits easier and less error-prone than having to worry about whether the half-type is an alias to float. I didn't think of adding them till working on this PR.